### ISSUE TITLE
Division

### ISSUE THREAD

**Issue by treeowl** [haskell-foundation/foundation#154]
The Prelude erred, I believe, in offering machine division (`quotRem`) and Knuthian division (`divMod`) as the standard division operations. I think we want to keep Knuthian division around for the situations where it's appropriate, but give a very nice name to "Euclidean" division, which guarantees a non-negative remainder.

**Comment by vincenthz** [comment]
I agree. My current thinking is that `Prelude.divMod` is very seldom useful (unless in specific domains maybe), and it would make sense to just hide this in a Math module [1], just like with the trigonometry stuff (`cos`,`sin`,etc that are moved to `Foundation.Math.Trigonometry`).

Also, I think `quot` and `rem` are not very good name to expose by default, and (much less strongly) I'm not sure about `div` and `mod` either.

* Any idea about naming all this ? I'm not that keen on `IDivisible` in the first place, but 
* Any alternative to expose the divisions ?

**Comment by snowleopard** [comment]
My opinion is that integer division `a/b` should follow the standard mathematical definition: if `b` is non-zero, then there are unique integers `q` (quotient) and `r` (remainder) such that `a = q * b + r` and `0 ≤ r < |d|`. 

I believe @treeowl refers to this as _Knuthian division_, but I'm not sure why -- as far as I know this is a commonly accepted definition in mathematics. Many programming languages get it wrong though.

Also `mod` is a standard mathematical notation.

**Comment by treeowl** [comment]
No, that's what I called Euclidean. Knuthian is the Prelude div/mod.

**Comment by snowleopard** [comment]
@treeowl Ah, I'm sorry, my mistake. To my embarrassment I didn't realise that Knuth's definition was different from the standard mathematical one, which apparently is referred to as Euclidean :-)

I fully support your proposal: the Euclidean division should be the default/easiest to write.

**Comment by vincenthz** [comment]
There's no division currently in prelude that guarantee a positive remainder. I think it would make sense to define use-case i.e. what does it mean when `a` is `X`, and `b` is `Y`.

@snowleopard: `mod` is a standard mathematical notation but is not defined `a -> a -> a`. more accurately it would be `a -> NonZero Natural -> a`.
